### PR TITLE
Add support for Druid's basic and kerberos authentication

### PR DESCRIPTION
**Pull request by beinan**
Basic authentication is supported by airlift's BasicAuthRequestFilter
Kerberos authentication is supported by the embedded SpnegoAuthentication in airlift's http-client

```
== RELEASE NOTES ==

General Changes
* Add support for Druid's basic and kerberos authentication
```

### CODE REVIEW

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java [hunk 1]
```diff
@@ -29,6 +29,14 @@
     private String schema = "druid";
     private boolean pushdown;
     private List<String> hadoopConfiguration = ImmutableList.of();
+    private DruidAuthenticationType druidAuthenticationType = DruidAuthenticationType.NONE;
+
+    public enum DruidAuthenticationType
+    {
+        NONE,
+        BASIC,
+        KERBEROS,
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java:41]: KERBEROS, -> KERBEROS

File: presto-druid/src/test/java/com/facebook/presto/druid/TestDruidConfig.java [hunk 2]
```diff
@@ -33,7 +33,8 @@ public void testDefaults()
                 .setDruidCoordinatorUrl(null)
                 .setDruidSchema("druid")
                 .setComputePushdownEnabled(false)
-                .setHadoopConfiguration(""));
+                .setHadoopConfiguration("")
+                .setDruidAuthenticationType(null));
```
Comment (zhenxiao) [inline presto-druid/src/test/java/com/facebook/presto/druid/TestDruidConfig.java]: should it be DruidAuthenticationType.NONE?

File: presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import static com.facebook.airlift.http.client.HttpClientBinder.httpClientBinder;
+
+public class DruidAuthenticationModule
+        extends AbstractConfigurationAwareModule
+{
+    @Override
+    protected void setup(Binder binder)
+    {
+        bindAuthenticationModule(
+                config -> config.getDruidAuthenticationType() == DruidConfig.DruidAuthenticationType.NONE,
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java:34]: static import DruidConfig.DruidAuthenticationType.NONE

File: presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                noneAuthenticationModule());
+
+        bindAuthenticationModule(
+                config -> config.getDruidAuthenticationType() == DruidConfig.DruidAuthenticationType.BASIC,
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java:38]: static import DruidConfig.DruidAuthenticationType.BASIC

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java [hunk 5]
```diff
@@ -30,6 +31,8 @@
     private boolean pushdown;
     private List<String> hadoopConfiguration = ImmutableList.of();
     private DruidAuthenticationType druidAuthenticationType = DruidAuthenticationType.NONE;
+    private String username;
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java:34]: s/username/basicAuthenticationUsername/g

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java [hunk 6]
```diff
@@ -30,6 +31,8 @@
...
+    private String password;
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java:35]: s/password/basicAuthenticationPassword/g

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java [hunk 7]
```diff
@@ -130,4 +133,30 @@ public DruidConfig setDruidAuthenticationType(DruidAuthenticationType druidAuthe
         }
         return this;
     }
+
+    @Nullable
+    public String getUsername()
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java:138]: getBasicAuthenticationUsername

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java [hunk 8]
```diff
@@ -130,4 +133,30 @@ public DruidConfig setDruidAuthenticationType(DruidAuthenticationType druidAuthe
...
+
+    @Config("druid.authentication.username")
+    public DruidConfig setUsername(String username)
+    {
+        this.username = username;
+        return this;
+    }
+
+    @Nullable
+    public String getPassword()
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java:151]: getBasicAuthenticationPassword

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java [hunk 9]
```diff
@@ -130,4 +133,30 @@ public DruidConfig setDruidAuthenticationType(DruidAuthenticationType druidAuthe
...
     }
+
+    @Nullable
+    public String getUsername()
+    {
+        return username;
+    }
+
+    @Config("druid.authentication.username")
+    public DruidConfig setUsername(String username)
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java:144]: druid.basic.authentication.username
setBasicAuthenticationUserName

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java [hunk 10]
```diff
@@ -130,4 +133,30 @@ public DruidConfig setDruidAuthenticationType(DruidAuthenticationType druidAuthe
...
+    }
+
+    @Nullable
+    public String getPassword()
+    {
+        return password;
+    }
+
+    @Config("druid.authentication.password")
+    public DruidConfig setPassword(String password)
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java:157]: similar to above

File: presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java [hunk 11]
```diff
@@ -37,6 +37,10 @@ protected void setup(Binder binder)
         bindAuthenticationModule(
                 config -> config.getDruidAuthenticationType() == DruidConfig.DruidAuthenticationType.BASIC,
                 basicAuthenticationModule());
+
+        bindAuthenticationModule(
+                config -> config.getDruidAuthenticationType() == DruidConfig.DruidAuthenticationType.KERBEROS,
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java:42]: static import DruidConfig.DruidAuthenticationType.KERBEROS

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java (hunk 1 above)
Comment (beinan) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java:41]: done

File: presto-druid/src/test/java/com/facebook/presto/druid/TestDruidConfig.java (hunk 2 above)
Comment (beinan) [reply presto-druid/src/test/java/com/facebook/presto/druid/TestDruidConfig.java]: changed to DruidAuthenticationType.NONE

File: presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java (hunk 3 above)
Comment (beinan) [reply presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java:34]: done

File: presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java (hunk 4 above)
Comment (beinan) [reply presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java:38]: done

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java (hunk 10 above)
Comment (beinan) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java:157]: done

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java (hunk 7 above)
Comment (beinan) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java]: done

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java (hunk 8 above)
Comment (beinan) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidConfig.java]: done

File: presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java (hunk 11 above)
Comment (beinan) [reply presto-druid/src/main/java/com/facebook/presto/druid/authentication/DruidAuthenticationModule.java]: done

File: presto-druid/src/test/java/com/facebook/presto/druid/TestDruidConfig.java [hunk 12]
```diff
@@ -33,7 +33,10 @@ public void testDefaults()
                 .setDruidCoordinatorUrl(null)
                 .setDruidSchema("druid")
                 .setComputePushdownEnabled(false)
-                .setHadoopConfiguration(""));
+                .setHadoopConfiguration("")
+                .setDruidAuthenticationType(DruidConfig.DruidAuthenticationType.NONE)
```
Comment (zhenxiao) [inline presto-druid/src/test/java/com/facebook/presto/druid/TestDruidConfig.java:37]: static import DruidConfig.DruidAuthenticationType.NONE?

Comment (beinan) [reply presto-druid/src/test/java/com/facebook/presto/druid/TestDruidConfig.java]: done

**Comment by beinan** [comment]
> hi @beinan looks good, just one small issue

Fixed, many thanks for review!

**Pull request merged**